### PR TITLE
feat: generic engineio socket

### PR DESCRIPTION
**Pull request by nshaaban-cPacket**
EngineIO Socket trait to allow for either future server engineio socket or client engineio socket in common socket.io socket

### CODE REVIEW

File: engineio/src/model.rs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+use crate::error::Result;
+use crate::Packet;
+
+pub trait Socket {
+    fn close(&self) -> Result<()>;
+    fn connect(&self) -> Result<()>;
+    fn emit(&self, packet: Packet) -> Result<()>;
+    fn is_connected(&self) -> Result<bool>;
```
Comment (1c3t3a) [inline engineio/src/model.rs:8]: Nice! Everything with `&_` allows to just pass an `&Socket` to the callback!

File: socketio/src/socket.rs [hunk 2]
```diff
@@ -41,8 +36,10 @@ pub struct Ack {
 
 /// Handles communication in the `socket.io` protocol.
 #[derive(Clone)]
-pub struct Socket {
-    engine_socket: Arc<EngineIoSocket>,
+pub struct Socket<
+    EngineSocket: rust_engineio::model::Socket + Send + Sync + Clone + 'static + Debug,
```
Comment (1c3t3a) [inline socketio/src/socket.rs:40]: Would it make sense to have those trait bounds directly on the Socket trait? I don't think there will be an implementation that's not Send + Sync + ...

**Review by 1c3t3a — COMMENTED** [review]
That looks really good! Very clean with the trait and the generic Type on the socket.

File: socketio/src/socket.rs (hunk 2 above)
Comment (nshaaban-cPacket) [reply socketio/src/socket.rs]: 86d501ac1fcf7116630e3e8895b69616349eb89d

File: engineio/src/model.rs (hunk 1 above)
Comment (nshaaban-cPacket) [reply engineio/src/model.rs:8]: EngineIO doesn't pass the socket to engineio callbacks today, only socketio exposes a socket in the callbacks.

**Comment by nshaaban-cPacket** [comment]
found a better way

**Pull request closed without merging**